core/server/common/src/main/java/alluxio/master/journal/raft/JournalStateMachine.java

```diff
@@ -138,8 +142,13 @@
    * @param journals      master journals; these journals are still owned by the caller, not by the
    *                      journal state machine
    * @param journalSystem the raft journal system
+   * @param maxConcurrencyPoolSize the maximum concurrency for notifyTermIndexUpdated
    */
-  public JournalStateMachine(Map<String, RaftJournal> journals, RaftJournalSystem journalSystem) {
+  public JournalStateMachine(Map<String, RaftJournal> journals, RaftJournalSystem journalSystem,
+                             Integer maxConcurrencyPoolSize) {
+    mJournalPool = Executors.newFixedThreadPool(maxConcurrencyPoolSize);
```

Comment:
It is generally preferable to use the ThreadPoolExecutor constructor which gives a lot more control over resources consumed. see https://www.baeldung.com/java-executors-cached-fixed-threadpool In this case, let's use a bounded task queue with some reasonable minimum thread count.

Reply:
I opted for a for 
```
ArrayBlockingQueue<Runnable> boundedQ = new ArrayBlockingQueue<>(maxConcurrencyPoolSize * 2);
mJournalPool = new ThreadPoolExecutor(maxConcurrencyPoolSize / 2, maxConcurrencyPoolSize, 0L,
        TimeUnit.MILLISECONDS, boundedQ);
```
The thread pool begins with half the max pool size, and can queue up to twice the max pool size.